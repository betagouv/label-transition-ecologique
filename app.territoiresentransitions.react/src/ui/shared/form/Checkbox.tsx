import classNames from 'classnames';

type TCheckbox = {
  label: string;
  onCheck: () => void;
  checked: boolean;
  disabled?: boolean;
  className?: string;
  labelClassName?: string;
};

const Checkbox = ({
  label,
  checked,
  onCheck,
  disabled,
  className,
  labelClassName,
}: TCheckbox) => {
  return (
    <div
      className={classNames(
        'fr-checkbox-group fr-checkbox-group--sm',
        className
      )}
    >
      <input
        type="checkbox"
        id="checkbox"
        name="checkbox"
        className="!w-4"
        onChange={onCheck}
        checked={checked}
        disabled={disabled}
      />
      <label
        className={classNames('fr-label', labelClassName)}
        htmlFor="checkbox"
      >
        {label}
      </label>
    </div>
  );
};

export default Checkbox;
