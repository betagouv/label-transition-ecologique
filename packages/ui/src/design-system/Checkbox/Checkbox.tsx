import classNames from 'classnames';
import {FieldMessage, FieldMessageProps} from '@design-system/Field';

export type CheckboxProps = Omit<
  React.ComponentPropsWithoutRef<'input'>,
  'type' | 'size'
> & {
  /** Libellé associé à la case à cocher */
  label?: string;
  /** Variante visuelle */
  variant?: 'checkbox' | 'switch';
} & FieldMessageProps;

/**
 * Affiche une case à cocher (ou un interrupteur) et optionnellement un
 * libellé, et un message complémentaire éventuellement stylé en fonction d'un
 * état.
 */
export const Checkbox = ({
  label,
  state,
  message,
  className,
  variant = 'checkbox',
  ...remainingProps
}: CheckboxProps) => {
  const input = (
    <input
      type="checkbox"
      className={classNames(
        // styles communs
        '!appearance-none checked:before:icon-check-line checked:bg-primary checked:disabled:bg-primary-5',
        {
          /** Checkbox */
          'flex content-center justify-center border border-solid !border-grey-6 rounded flex-wrap w-5 h-5 !text-white  checked:!border-transparent checked:hover:bg-primary-8 checked:!disabled:border-grey-4':
            variant === 'checkbox',

          /** Switch */
          'relative bg-grey-4 rounded-full w-10 h-6 before:absolute before:h-4 before:w-4 before:left-1 before:top-1 before:bg-white before:duration-200 before:rounded-full checked:text-primary checked:before:translate-x-4 checked:before:flex checked:before:items-center checked:before:justify-center':
            variant === 'switch',
        },

        className
      )}
      {...remainingProps}
    />
  );

  return (
    <div className="flex flex-col w-max">
      {label ? (
        /** affiche l'input et le libellé */
        <label
          className={classNames('inline-flex items-center cursor-pointer ', {
            'text-grey-8': !remainingProps.disabled,
            'text-grey-6': remainingProps.disabled,
          })}
        >
          {input}
          <span className="pl-2">{label}</span>
        </label>
      ) : (
        /** OU affiche que l'input si le libellé n'est pas renseigné */
        input
      )}
      {/** Message complémentaire */}
      <FieldMessage
        state={state}
        message={message}
        messageClassName={classNames('pt-1', {
          'pl-7': variant === 'checkbox',
          'pl-12': variant === 'switch',
        })}
      />
    </div>
  );
};
