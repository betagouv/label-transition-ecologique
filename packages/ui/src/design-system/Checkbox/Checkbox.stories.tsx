import {useState} from 'react';
import {Meta, StoryObj} from '@storybook/react';
import {Checkbox} from './Checkbox';

const meta: Meta<typeof Checkbox> = {
  title: 'Design System/Checkbox',
  component: Checkbox,
  render: args => {
    const [checked, setChecked] = useState(args.checked || false);
    return (
      <Checkbox
        {...args}
        checked={checked}
        label="Description de l'action"
        onChange={() => setChecked(!checked)}
      />
    );
  },
};

export default meta;

type Story = StoryObj<typeof Checkbox>;

export const SansLabel: Story = {
  args: {},
  render: args => {
    const [checked, setChecked] = useState(args.checked || false);
    return (
      <Checkbox
        {...args}
        checked={checked}
        onChange={() => setChecked(!checked)}
      />
    );
  },
};

export const Unchecked: Story = {
  args: {},
};

export const Checked: Story = {
  args: {
    checked: true,
  },
};

export const UncheckedAndDisabled: Story = {
  args: {
    checked: false,
    disabled: true,
  },
};

export const CheckedAndDisabled: Story = {
  args: {
    checked: true,
    disabled: true,
  },
};

export const AvecMessage: Story = {
  args: {
    message: 'Description additionnelle',
  },
};

export const AvecMessageEtEtat: Story = {
  args: {
    message: 'Description additionnelle',
    state: 'info',
  },
};

export const VarianteSwitch: Story = {
  args: {
    message: 'Description additionnelle',
    state: 'info',
    variant: 'switch',
  },
};

export const VarianteSwitchChecked: Story = {
  args: {
    checked: true,
    message:
      'Le lorem ipsum est, en imprimerie, une suite de mots sans signification utilis√©e',
    state: 'info',
    variant: 'switch',
  },
};
