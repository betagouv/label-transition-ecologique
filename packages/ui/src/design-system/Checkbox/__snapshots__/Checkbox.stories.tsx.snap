// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Design System/Checkbox AvecMessage smoke-test 1`] = `
<div class="flex gap-2 min-w-min">
  <input id="cb7"
         type="checkbox"
         class="!appearance-none shrink-0 checked:before:icon-check-line checked:bg-primary checked:disabled:bg-primary-5 flex content-center justify-center mt-0.5 border border-solid !border-grey-6 rounded flex-wrap w-5 h-5 !text-white  checked:!border-transparent checked:hover:bg-primary-8 checked:!disabled:border-grey-4"
  >
  <div class="flex flex-col gap-1">
    <label for="cb7"
           class="inline-flex items-center cursor-pointer  text-grey-8"
    >
      Description de l'action
    </label>
    <div class="flex items-center text-grey-7 my-auto">
      <span class="text-xs">
        Description additionnelle
      </span>
    </div>
  </div>
</div>
`;

exports[`Design System/Checkbox AvecMessageEtEtat smoke-test 1`] = `
<div class="flex gap-2 min-w-min">
  <input id="cb8"
         type="checkbox"
         class="!appearance-none shrink-0 checked:before:icon-check-line checked:bg-primary checked:disabled:bg-primary-5 flex content-center justify-center mt-0.5 border border-solid !border-grey-6 rounded flex-wrap w-5 h-5 !text-white  checked:!border-transparent checked:hover:bg-primary-8 checked:!disabled:border-grey-4"
  >
  <div class="flex flex-col gap-1">
    <label for="cb8"
           class="inline-flex items-center cursor-pointer  text-grey-8"
    >
      Description de l'action
    </label>
    <div class="flex items-center text-info-1 my-auto">
      <span class="ri-information-fill font-normal text-center h-4 w-4 text-[1rem] leading-[1rem] mr-1">
      </span>
      <span class="text-xs">
        Description additionnelle
      </span>
    </div>
  </div>
</div>
`;

exports[`Design System/Checkbox AvecMessageUniquement smoke-test 1`] = `
<div class="flex gap-2 min-w-min">
  <input id="cb9"
         type="checkbox"
         class="!appearance-none shrink-0 checked:before:icon-check-line checked:bg-primary checked:disabled:bg-primary-5 flex content-center justify-center mt-0.5 border border-solid !border-grey-6 rounded flex-wrap w-5 h-5 !text-white  checked:!border-transparent checked:hover:bg-primary-8 checked:!disabled:border-grey-4"
  >
  <div class="flex flex-col gap-1">
    <div class="flex items-center text-grey-7 my-auto">
      <span class="text-xs">
        Le lorem ipsum est, en imprimerie, une suite de mots sans signification utilisée
      </span>
    </div>
  </div>
</div>
`;

exports[`Design System/Checkbox Checked smoke-test 1`] = `
<div class="flex gap-2 min-w-min">
  <input id="cb3"
         type="checkbox"
         class="!appearance-none shrink-0 checked:before:icon-check-line checked:bg-primary checked:disabled:bg-primary-5 flex content-center justify-center mt-0.5 border border-solid !border-grey-6 rounded flex-wrap w-5 h-5 !text-white  checked:!border-transparent checked:hover:bg-primary-8 checked:!disabled:border-grey-4"
         checked
  >
  <div class="flex flex-col gap-1">
    <label for="cb3"
           class="inline-flex items-center cursor-pointer  text-grey-8"
    >
      Description de l'action
    </label>
  </div>
</div>
`;

exports[`Design System/Checkbox CheckedAndDisabled smoke-test 1`] = `
<div class="flex gap-2 min-w-min">
  <input id="cb5"
         type="checkbox"
         class="!appearance-none shrink-0 checked:before:icon-check-line checked:bg-primary checked:disabled:bg-primary-5 flex content-center justify-center mt-0.5 border border-solid !border-grey-6 rounded flex-wrap w-5 h-5 !text-white  checked:!border-transparent checked:hover:bg-primary-8 checked:!disabled:border-grey-4"
         disabled
         checked
  >
  <div class="flex flex-col gap-1">
    <label for="cb5"
           class="inline-flex items-center cursor-pointer  text-grey-6"
    >
      Description de l'action
    </label>
  </div>
</div>
`;

exports[`Design System/Checkbox LongLabelAvecMessage smoke-test 1`] = `
<div class="max-w-[16rem]">
  <div class="flex gap-2 min-w-min">
    <input id="cb6"
           type="checkbox"
           class="!appearance-none shrink-0 checked:before:icon-check-line checked:bg-primary checked:disabled:bg-primary-5 flex content-center justify-center mt-0.5 border border-solid !border-grey-6 rounded flex-wrap w-5 h-5 !text-white  checked:!border-transparent checked:hover:bg-primary-8 checked:!disabled:border-grey-4"
    >
    <div class="flex flex-col gap-1">
      <label for="cb6"
             class="inline-flex items-center cursor-pointer  text-grey-8"
      >
        Lorem ipsum dolor sit amet, consectetur adipiscing elit.
      </label>
      <div class="flex items-center text-grey-7 my-auto">
        <span class="text-xs">
          Le lorem ipsum est, en imprimerie, une suite de mots sans signification utilisée
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`Design System/Checkbox SansLabel smoke-test 1`] = `
<div class="flex gap-2 min-w-min">
  <input id="cb1"
         type="checkbox"
         class="!appearance-none shrink-0 checked:before:icon-check-line checked:bg-primary checked:disabled:bg-primary-5 flex content-center justify-center mt-0.5 border border-solid !border-grey-6 rounded flex-wrap w-5 h-5 !text-white  checked:!border-transparent checked:hover:bg-primary-8 checked:!disabled:border-grey-4"
  >
  <div class="flex flex-col gap-1">
  </div>
</div>
`;

exports[`Design System/Checkbox Unchecked smoke-test 1`] = `
<div class="flex gap-2 min-w-min">
  <input id="cb2"
         type="checkbox"
         class="!appearance-none shrink-0 checked:before:icon-check-line checked:bg-primary checked:disabled:bg-primary-5 flex content-center justify-center mt-0.5 border border-solid !border-grey-6 rounded flex-wrap w-5 h-5 !text-white  checked:!border-transparent checked:hover:bg-primary-8 checked:!disabled:border-grey-4"
  >
  <div class="flex flex-col gap-1">
    <label for="cb2"
           class="inline-flex items-center cursor-pointer  text-grey-8"
    >
      Description de l'action
    </label>
  </div>
</div>
`;

exports[`Design System/Checkbox UncheckedAndDisabled smoke-test 1`] = `
<div class="flex gap-2 min-w-min">
  <input id="cb4"
         type="checkbox"
         class="!appearance-none shrink-0 checked:before:icon-check-line checked:bg-primary checked:disabled:bg-primary-5 flex content-center justify-center mt-0.5 border border-solid !border-grey-6 rounded flex-wrap w-5 h-5 !text-white  checked:!border-transparent checked:hover:bg-primary-8 checked:!disabled:border-grey-4"
         disabled
  >
  <div class="flex flex-col gap-1">
    <label for="cb4"
           class="inline-flex items-center cursor-pointer  text-grey-6"
    >
      Description de l'action
    </label>
  </div>
</div>
`;

exports[`Design System/Checkbox VarianteSwitch smoke-test 1`] = `
<div class="flex gap-2 min-w-min">
  <input id="cb10"
         type="checkbox"
         class="!appearance-none shrink-0 checked:before:icon-check-line checked:bg-primary checked:disabled:bg-primary-5 relative bg-grey-4 rounded-full w-10 h-6 before:absolute before:h-4 before:w-4 before:left-1 before:top-1 before:bg-white before:duration-200 before:rounded-full checked:text-primary checked:before:translate-x-4 checked:before:flex checked:before:items-center checked:before:justify-center"
  >
  <div class="flex flex-col gap-1">
    <label for="cb10"
           class="inline-flex items-center cursor-pointer  text-grey-8"
    >
      Description de l'action
    </label>
    <div class="flex items-center text-info-1 my-auto">
      <span class="ri-information-fill font-normal text-center h-4 w-4 text-[1rem] leading-[1rem] mr-1">
      </span>
      <span class="text-xs">
        Description additionnelle
      </span>
    </div>
  </div>
</div>
`;

exports[`Design System/Checkbox VarianteSwitchChecked smoke-test 1`] = `
<div class="flex gap-2 min-w-min">
  <input id="cb11"
         type="checkbox"
         class="!appearance-none shrink-0 checked:before:icon-check-line checked:bg-primary checked:disabled:bg-primary-5 relative bg-grey-4 rounded-full w-10 h-6 before:absolute before:h-4 before:w-4 before:left-1 before:top-1 before:bg-white before:duration-200 before:rounded-full checked:text-primary checked:before:translate-x-4 checked:before:flex checked:before:items-center checked:before:justify-center"
         checked
  >
  <div class="flex flex-col gap-1">
    <label for="cb11"
           class="inline-flex items-center cursor-pointer  text-grey-8"
    >
      Description de l'action
    </label>
    <div class="flex items-center text-info-1 my-auto">
      <span class="ri-information-fill font-normal text-center h-4 w-4 text-[1rem] leading-[1rem] mr-1">
      </span>
      <span class="text-xs">
        Le lorem ipsum est, en imprimerie, une suite de mots sans signification utilisée
      </span>
    </div>
  </div>
</div>
`;

exports[`Design System/Checkbox VarianteSwitchDisabled smoke-test 1`] = `
<div class="flex gap-2 min-w-min">
  <input id="cb12"
         type="checkbox"
         class="!appearance-none shrink-0 checked:before:icon-check-line checked:bg-primary checked:disabled:bg-primary-5 relative bg-grey-4 rounded-full w-10 h-6 before:absolute before:h-4 before:w-4 before:left-1 before:top-1 before:bg-white before:duration-200 before:rounded-full checked:text-primary checked:before:translate-x-4 checked:before:flex checked:before:items-center checked:before:justify-center"
         disabled
         checked
  >
  <div class="flex flex-col gap-1">
    <label for="cb12"
           class="inline-flex items-center cursor-pointer  text-grey-6"
    >
      Description de l'action
    </label>
  </div>
</div>
`;
